### PR TITLE
cabana: add filter per column

### PR DESCRIPTION
**Pull request by pd0wm**
Screenshot 2023-05-04 at 10 33 22
Screenshot 2023-05-04 at 10 33 37

- [x] Change top header to contain msg/signal count, clear all filters and multiple lines checkbox
- [x] Test behavior during horizontal scrolling
- [x] Test behavior with hidden columns
- [x] Add range/wildcard filtering for numbers (e.g. 100-200, 100-)
- [x] Add regex filter
- [x] Re-evaluate filters when new data comes in (needed for count, frequency and data)

### CODE REVIEW

File: tools/cabana/messageswidget.cc [hunk 1]
```diff
@@ -47,21 +38,26 @@ MessagesWidget::MessagesWidget(QWidget *parent) : QWidget(parent) {
...
 
   // signals/slots
-  QObject::connect(filter, &QLineEdit::textEdited, model, &MessageListModel::setFilterString);
+  QObject::connect(header, &MessageViewHeader::filtersUpdated, model, &MessageListModel::setFilterStrings);
   QObject::connect(multiple_lines_bytes, &QCheckBox::stateChanged, [=](int state) {
     settings.multiple_lines_bytes = (state == Qt::Checked);
     delegate->setMultipleLines(settings.multiple_lines_bytes);
     view->setUniformRowHeights(!settings.multiple_lines_bytes);
-    model->sortMessages();
+    model->fetchData(); // Why?
```
Comment (deanlee) [inline tools/cabana/messageswidget.cc:54]: The reason I call `sortMessages` here is to reset the model to force the view to recalculate the size of the bytes column. 
This call should be necessary for messages that have not been updated by updateState（or cabana is in pause mode)

Comment (pd0wm) [reply tools/cabana/messageswidget.cc:54]: Shall we just call begin/endresetModel then?

Comment (deanlee) [reply tools/cabana/messageswidget.cc:54]: Yes, it's fine to just reset the model here . This way the view will recalculate the size and  display bytes column correctly.

Comment (pd0wm) [reply tools/cabana/messageswidget.cc]: Do I need to use a CAN FD route for multiline bytes? Is there a public one you've been using for testing?

Comment (deanlee) [reply tools/cabana/messageswidget.cc]: Frankly, I just added some temporary code to `MessageListModel::data` for testing:

```
 } else if (role == BytesRole && index.column() == 5) {
    if (QSet<int>({1, 3, 6, 7}).contains(index.row())) {
      QByteArray dat(64, 0);
      memcpy(dat.data(), can_data.dat.data(), can_data.dat.size());
      return dat;
    }
    return can_data.dat;
  }
```

**Comment by deanlee** [comment]
There doesn't seem to be a simple way to reset filters. would it be better to place a "Reset Filters" button above the header?

**Comment by pd0wm** [comment]
> 

Sounds good. I was looking to add the number of signals somewhere as well, since we like that too as a metric of RE progress.

**Comment by pd0wm** [comment]
@adeebshihadeh This is ready for review

@deanlee Happy with the new logic on when to call `resetModel()`? `fetchData` always filters and sorts the messages, and then compares if the list of IDs is different, if so it initiates the reset.

**Comment by deanlee** [comment]
It's very nice. I'll implement the features we discussed in https://github.com/commaai/openpilot/pull/28093#issuecomment-1534215763  after this is merged, with some minor optimizations if necessary.

**Pull request merged**